File: src/kernelprolog/fluents/SxxMachine/SourceMerger.cpp
using namespace std;

#include "SourceMerger.h"
#include "../../../machine/SxxMachine/Nonvar.h"
#include "../../terms/SxxMachine/Prog.h"
#include "../../terms/SxxMachine/Copier.h"
#include "O1Queue.h"
#include "../../../machine/SxxMachine/Term.h"
#include "../../terms/SxxMachine/Source.h"

namespace SxxMachine
{

	SourceMerger::SourceMerger(Nonvar *Xs, Prog *p) : IterableSource(p)
	{
		this->Q = new O1Queue(Copier::ConsToVector(Xs));
	}

	Term *SourceMerger::getElement()
	{
		if (nullptr == Q)
		{
			return nullptr;
		}
		while (!Q->isEmpty())
		{
			Source *current = any_cast<Source*>(Q->deq());
			if (nullptr == current)
			{
				continue;
			}
			Term *T = current->getElement();
			if (nullptr == T)
			{
				continue;
			}
			Q->enq(current);
			return T;
		}
		return nullptr;
	}
}

File: jsrc-kernelprolog/fluents/SxxMachine/SourceMerger.java
package SxxMachine;

import SxxMachine.pterm.SourceFluentTerm;

/**
 * Merges a List of Sources into a new Source which (fairly) iterates over them
 * breadth first.
 */
public class SourceMerger extends IterableSource {

    public SourceMerger(Nonvar Xs, Prog p) {
        super(p);
        this.Q = new O1Queue(Copier.ConsToVector(Xs));
    }

    private O1Queue Q;

    @Override
    public Term getElement() {
        if (null == Q)
            return null;
        while (!Q.isEmpty()) {
            SourceFluentTerm current = (SourceFluentTerm) Q.deq();
            if (null == current)
                continue;
            Term T = current.getElement();
            if (null == T)
                continue;
            Q.enq(current);
            return T;
        }
        return null;
    }
}
